cmd/snap-install/volmgr/volmgr.go

```diff
@@ -0,0 +1,216 @@
+	if err != nil {
+		return nil, err
+	}
+
+	constraints := gadget.LayoutConstraints{
+		NonMBRStartOffset: 1 * gadget.SizeMiB,
+		SectorSize:        512,
+	}
+
+	positionedVolume := map[string]*gadget.LaidOutVolume{}
+
+	for name, vol := range info.Volumes {
+		pvol, err := gadget.LayoutVolume(gadgetRoot, &vol, constraints)
+		if err != nil {
+			return nil, err
+		}
+		positionedVolume[name] = pvol
+	}
+
+	v := &VolumeManager{
+		gadgetRoot:       gadgetRoot,
+		device:           device,
+		partitionTool:    newPartitionTool(device),
+		info:             info,
+		positionedVolume: positionedVolume,
+	}
+
+	return v, nil
+}
+
+func (v *VolumeManager) Run() error {
+
+	if err := v.readDevice(); err != nil {
+		return err
+	}
+
+	if err := v.completeLayout(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// readDevice reads the partition layout from the block device.
+func (v *VolumeManager) readDevice() error {
+
+	var err error
+	if v.deviceVolume, err = v.partitionTool.deviceInfo(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// completeLayout adds missing structures from the gadget specification.
+func (v *VolumeManager) completeLayout() error {
+
+	// Limit ourselves to just one volume for now.
+	if len(v.positionedVolume) != 1 {
+		return fmt.Errorf("multiple volumes are not supported")
+	}
+	var name string
+	for k := range v.positionedVolume {
+		name = k
+	}
+	gadgetVolume := v.positionedVolume[name]
+
+	// Check if all existing device partitions are also in gadget
+	usedPartitions := make([]bool, len(gadgetVolume.LaidOutStructure))
+	for _, ps := range v.deviceVolume.LaidOutStructure {
+		index, err := findStructureInVolume(&ps, gadgetVolume)
+		if err != nil {
+			return err
+		}
+		if index >= len(usedPartitions) {
+			return fmt.Errorf("gadget structure indexes are inconsistent")
+		}
+		usedPartitions[index] = true
+	}
+
+	// Map device nodes to structure names
+	deviceMap := map[string]string{}
+
+	// Create missing partitions
+	if err := v.partitionTool.createPartitions(gadgetVolume, usedPartitions, deviceMap); err != nil {
+		return err
+	}
+
+	// Make filesystems on the newly created partitions
+	for i, p := range gadgetVolume.LaidOutStructure {
+		s := p.VolumeStructure
+		// Skip partitions that are already in the volume
+		if usedPartitions[i] {
+			continue
+		}
+		// Skip MBR structure
+		if s.Type == "mbr" {
+			continue
+		}
+
+		node, ok := deviceMap[s.Name]
+		if !ok {
+			continue
+		}
+		if err := makeFilesystem(node, s.Label, s.Filesystem); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func makeFilesystem(node, label, filesystem string) error {
+	switch filesystem {
+	case "vfat":
+		return makeVFATFilesystem(node, label)
+	default:
+		return fmt.Errorf("cannot create unsupported filesystem %q", filesystem)
+	}
+}
+
+func makeVFATFilesystem(node, label string) error {
+	output, err := exec.Command("mkfs.vfat", "-n", label, node).CombinedOutput()
+	if err != nil {
```

Comment:
I think I did that to prevent very long lines in the ext4 case and left vfat as that to keep it consistent, but now that I added ext4 I see that it isn't that much longer. Using the compact version instead.